### PR TITLE
Added location field to project

### LINKED ISSUE #287: Project date and location should not be a text field
Location can remain a text field, processed like on Hasjob to identify location tags. The display date should be auto-calculated from the existing start and end fields. The existing date and location field needs to be (a) renamed or (b) deleted. Either case will involve manually editing the value for the existing 150+ projects.

### PR DESCRIPTION
**Pull request by iambibhas**
Partially Fixes #287. Going to do the data migration with an ad hoc script.

### CODE REVIEW

File: migrations/versions/92e1cc95c7f7_project_location_field.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
+"""project location field
+
+Revision ID: 92e1cc95c7f7
+Revises: 2441cb4f44d4
+Create Date: 2018-11-24 08:27:06.710711
+
+"""
+
+revision = '92e1cc95c7f7'
+down_revision = '2441cb4f44d4'
```
Comment (iambibhas) [inline migrations/versions/92e1cc95c7f7_project_location_field.py]: This will change once the previous PR about primary venue is merged.

**Comment by iambibhas** [comment]
The script to update date, date_upto and location from existing data - https://gist.github.com/iambibhas/a2f2e94dcbd48da6c1be6d17b7afb303#file-datelocation-sql

File: funnel/models/project.py [hunk 2]
```diff
@@ -44,9 +44,13 @@ class Project(BaseScopedNameMixin, db.Model):
     tagline = db.Column(db.Unicode(250), nullable=False)
     description = MarkdownColumn('description', default=u'', nullable=False)
     instructions = MarkdownColumn('instructions', default=u'', nullable=True)
+
     datelocation = db.Column(db.Unicode(50), default=u'', nullable=False)
+    location = db.Column(db.Unicode(50), default=u'', nullable=True)
```
Comment (jace) [inline funnel/models/project.py]: Just occurred to me that this conflicts with the Venue field. We may not even need a location field.

Comment (jace) [reply funnel/models/project.py]: Also, why 50 chars?

**Comment by jace** [comment]
This will also need a `parsed_location` JSON field, plus a separate `ProjectLocation` model to handle multi-location projects. See Hasjob's implementation.

File: funnel/models/project.py [hunk 3]
```diff
@@ -448,3 +453,17 @@ def migrate_profile(cls, oldprofile, newprofile):
...
         oldprofile.project_redirects = []
         return [cls.__table__.name]
+
+
+class ProjectLocation(TimestampMixin, db.Model):
+    __tablename__ = 'project_location'
+    #: Job post we are tagging
+    project_id = db.Column(None, db.ForeignKey('project.id'), primary_key=True, nullable=False)
+    project = db.relationship(Project, backref=db.backref('locations', cascade='all, delete-orphan'))
+    #: Geonameid for this job post
```
Comment (jace) [inline funnel/models/project.py]: Not a job post.

File: funnel/jobs/jobs.py [hunk 4]
```diff
@@ -15,3 +18,51 @@ def import_tickets(ticket_client_id):
                 ticket_list = Boxoffice(access_token=ticket_client.client_access_token).get_tickets(ticket_client.client_eventid)
                 ticket_client.import_from_list(ticket_list)
             db.session.commit()
+
+
+def tag_locations(project_id):
```
[the remark this answers is not in the document]
Comment (jace) [reply funnel/jobs/jobs.py]: In the view when project details are edited. This needs an RQ `@job` decorator.

File: funnel/extapi/boxoffice.py [hunk 5]
```diff
@@ -21,7 +21,11 @@ def __init__(self, access_token, base_url=None):
 
     def get_orders(self, ic):
         url = urljoin(self.base_url, 'ic/{ic}/orders?access_token={token}'.format(ic=ic, token=self.access_token))
-        return requests.get(url).json().get('orders')
+        resp = requests.get(url)
+        if resp.status_code == 200:
+            return resp.json().get('orders')
+        else:
+            raise Exception(u"Invalid response from boxoffice server: {0!r}".format(resp))
```
Comment (jace) [inline funnel/extapi/boxoffice.py]: Why do you need this? The `.json()` call will raise an exception anyway.

Comment (jace) [reply funnel/extapi/boxoffice.py]: In general, I think raising `Exception` is lazy and indicates poor planning. There has to be a more suitable exception.

File: funnel/views/project.py [hunk 6]
```diff
@@ -128,13 +130,15 @@ def project_edit(profile, project):
         form = ProposalSubprojectForm(obj=project, model=Project)
     else:
         form = ProjectForm(obj=project, model=Project)
+    del form.name
```
Comment (jace) [inline funnel/views/project.py]: Remove this. We'll do a clean fix for editing `name`, not this hack.

Comment (iambibhas) [reply funnel/views/project.py]: Fixed the form validation for name.

File: funnel/forms/project.py [hunk 7]
```diff
@@ -63,8 +63,14 @@ class ProjectForm(forms.Form):
         description=__(u"Eg: Explara, Instamojo"),
         validators=[forms.validators.Optional(), forms.validators.Length(max=2000)])
 
-    def validate_date_upto(self, date_upto):
-        if self.date_upto.data < self.date.data:
+    def validate_name(self, field):
+        if field.data != field.object_data:
+            # field.object_data returns the field data from the object attached to it.
+            # When a new project is being created, field.object_data is empty.
+            AvailableName()(self, field)
```
Comment (jace) [inline funnel/forms/project.py]: You don't need this. AvailableName will do this for you if the form has `edit_obj`, `edit_parent` and `edit_model`.

File: funnel/forms/project.py [hunk 8]
```diff
@@ -63,8 +63,8 @@ class ProjectForm(forms.Form):
         description=__(u"Eg: Explara, Instamojo"),
         validators=[forms.validators.Optional(), forms.validators.Length(max=2000)])
 
-    def validate_date_upto(self, date_upto):
-        if self.date_upto.data < self.date.data:
+    def validate_date_upto(self, field):
+        if field.data < self.date.data:
             raise forms.ValidationError(_("End date cannot be before start date"))
```
Comment (jace) [inline funnel/forms/project.py]: Use the `GreaterThan` validator. https://github.com/hasgeek/baseframe/blob/master/baseframe/forms/validators.py#L80-L94

Comment (jace) [reply funnel/forms/project.py]: Sorry, `LesserThan` here.

Comment (iambibhas) [reply funnel/forms/project.py]: Validator on `date_upto` will be `GreaterThan('date')`

**Pull request merged**